### PR TITLE
i/6053: The gravity of the selection should be overridden if a link was pasted for better typing experience

### LINKED ISSUE #6053: Selection could be outside of the link after pasting
## 📝 Provide a description of the improvement

*How the feature works now and what you'd like to change*?

When pasting a link the selection (cursor) could be moved outside of the link, so after pasting typing won't extend the pasted link. This behavior doesn't exist in CKEditor 4 and other editors.

## 📃 Other details

* Browser: Chrome
* OS: MacOS
* CKEditor version: 16

---

If you'd like to see this improvement implemented, add a 👍 reaction to this post.

### PR DESCRIPTION
**Pull request by oleq**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Fix: The gravity of the selection should be overridden if a link was pasted for better typing experience. Closes ckeditor/ckeditor5#6053.

### CODE REVIEW

File: src/linkediting.js [hunk 1]
```diff
@@ -250,4 +253,116 @@ export default class LinkEditing extends Plugin {
...
+				return;
+			}
+
+			// Filter out the following case where the selection has the "linkHref" attribute because the
+			// gravity is already overridden and some text with another attribute (e.g. <b>PASTED</b>) is pasted:
+			//
+			// Before paste:
+			//
+			//		 ↱
+			//		[]<$text linkHref="foo">link</$text>
```
Comment (jodator) [inline src/linkediting.js:314]: What should happen in a case below:

1.  Go inside link.
2.  Paste bolded text.
3.  Type.

What happens is that you type with but bold wasn't pasted):

File: src/linkediting.js [hunk 2]
```diff
@@ -250,4 +253,116 @@ export default class LinkEditing extends Plugin {
...
+			//		                                  ↱
+			//		<$text bold="true">PASTED</$text>[]<$text linkHref="foo">link</$text>
+			//
+			if ( !nodeBefore.hasAttribute( 'linkHref' ) ) {
+				return;
+			}
+
+			// Filter out the following case where a link is a pasted in the middle (or before) another link
+			// (different URLs, so they will not merge). In this (let's say weird) case, we can leave the gravity
+			// as it is because it would stick to the first link anyway.
```
Comment (jodator) [inline src/linkediting.js:327]: I understood this as "it would stick to the first link ('foo') anyway" but the behaviour is that it sticks to the pasted link (also as shown on the sample below). I'd rephrase that comment to remove ambiguity.

**Comment by jodator** [comment]
@oleq I've found something like this:



1.  Paste link at the end of a text.
2.  Move cursor to highlight link using two-step caret movement.
3.  Start typing.

Expected: You should type with link.

Actual: Only the first character is typed with link. Subsequent are plain text.

Additional info: if you change selection to other place in the editor (other link, typing) and go back to the same position as in tested scenario the type with link works as expected.

Looks like minor issue to me if not something easy to spot.

Edit:

After step 1:



After step 2 (gravity is overridden):



After moving selection after other link (gravity is not overridden):

File: src/linkediting.js [hunk 3]
```diff
@@ -250,4 +253,116 @@ export default class LinkEditing extends Plugin {
...
+
+		// Restore the selection gravity automatically as soon as the selection range changes.
+		selection.on( 'change:range', ( evt, data ) => {
+			// Don't restore the gravity if it wasn't overridden in the first place.
+			if ( !gravityOverrideUid ) {
+				return;
+			}
+
+			// Skip automatic restore when the change is indirect.
+			// It means that e.g. if the change was external (collaboration) and the user had their
```
Comment (jodator) [inline src/linkediting.js:356]: Needs an ending ;)

File: src/linkediting.js [hunk 4]
```diff
@@ -250,4 +253,116 @@ export default class LinkEditing extends Plugin {
...
+				return;
+			}
+
+			model.change( writer => {
+				gravityOverrideUid = writer.overrideSelectionGravity();
+			} );
+		}, { priority: 'low' } );
+
+		// Restore the selection gravity automatically as soon as the selection range changes.
+		selection.on( 'change:range', ( evt, data ) => {
```
Comment (jodator) [inline src/linkediting.js:349]: This doesn't work in a case from a comment. The gravity change does not trigger this fixer (the ranges doesn't change) so you can't type with link.

File: tests/linkediting.js [hunk 5]
```diff
@@ -19,22 +21,34 @@ import { keyCodes } from '@ckeditor/ckeditor5-utils/src/keyboard';
...
+					view: 'b'
+				} );
+
 				model = editor.model;
 				view = editor.editing.view;
 			} );
 	} );
 
 	afterEach( () => {
 		editor.destroy();
```
Comment (jodator) [inline tests/linkediting.js:50]: Would be safer to wait for a promise to resolve.

I'm using `async/await` in test, but can be done with removing an element and then returning the `editor.destroy()` promise.

Comment (jodator) [inline tests/linkediting.js:50]: Would be safer to wait for a promise to resolve.

I'm using `async/await` in test, but can be done with removing an element and then returning the `editor.destroy()` promise.

File: tests/linkediting.js [hunk 6]
```diff
@@ -73,8 +87,8 @@ describe( 'LinkEditing', () => {
 		} );
 
 		it( 'should be bound to th `linkHref` attribute (RTL)', () => {
-			return VirtualTestEditor
-				.create( {
+			return ClassicTestEditor
```
Comment (jodator) [inline tests/linkediting.js:90]: Might be a follow-up but AFAICS the main `beforeEach()` editor was not destroyed and this looks like we could have two editors there.

Comment (jodator) [reply tests/linkediting.js:90]: Adding `element` in tests and using `ClassicTestEditor` might create some memory leaks if multiple editors are created without removing previous one (from global `beforeEach()` ).

File: tests/linkediting.js [hunk 7]
```diff
@@ -104,6 +118,189 @@ describe( 'LinkEditing', () => {
...
+				'<paragraph>' +
+					'<$text linkHref="http://PASTED">PASTED[]</$text>' +
+					'<$text linkHref="ckeditor.com">foo</$text>' +
+				'</paragraph>'
+			);
+
+			expect( model.document.selection.isGravityOverridden ).to.be.false;
+		} );
+
+		it( 'should restore gravity on the next change:range', () => {
```
Comment (jodator) [inline tests/linkediting.js:249]: Looks like the test description is too technical ;) and will need to change. I'd go with the expected behaviour like when user move selection, or something similar.

File: tests/linkediting.js [hunk 8]
```diff
@@ -104,6 +118,189 @@ describe( 'LinkEditing', () => {
...
+		it( 'should not restore gravity if it wasn\'t overridden in the first place', () => {
+			const spy = sinon.spy( Writer.prototype, 'restoreSelectionGravity' );
+
+			setModelData( model, '<paragraph>fo[]o</paragraph>' );
+			model.modifySelection( model.document.selection );
+
+			sinon.assert.notCalled( spy );
+		} );
+
+		it( 'should not restore the gravity if the change:range was indirect (e.g. comming from the collaboration)', () => {
```
Comment (jodator) [inline tests/linkediting.js:280]:
```suggestion
		it( 'should not restore the gravity if the selection change was indirect (e.g. comming from the collaboration)', () => {
```

File: tests/linkediting.js [hunk 9]
```diff
@@ -625,9 +824,20 @@ describe( 'LinkEditing', () => {
 		} );
 
 		describe( 'upcast converter', () => {
+			let element, editor;
+
+			beforeEach( () => {
```
Comment (jodator) [inline tests/linkediting.js:829]: Another nested `beforeEach` might be probably safe as it handles different editors. I don't like that there are two editors on the same page in this tests.



Again, this does not break now, but we need at least a follow-up for this whole file.

**Review by jodator — CHANGES_REQUESTED** [review]
I've found one behaviour inconsistency - looks like listening on `change:range` event is too narrow.

**Comment by oleq** [comment]
FYI we went with a better approach https://github.com/ckeditor/ckeditor5-link/pull/261.
